cf-custom-resources/lib/custom-domain-app-runner.js

```diff
@@ -0,0 +1,189 @@
+        });
+    });
+
+    if (!addCustomDomainErr) {
+        console.log("Finished");
+        await report(event, context, "SUCCESS", event.LogicalResourceId).catch((err) => {
+            throw new Error("send response: " + err.message);
+        });
+    }
+};
+
+/**
+ * Validate certificates of the custom domain for the service by upserting validation records.
+ * Errors are not handled and are directly passed to the caller.
+ *
+ * @param {string} serviceARN ARN of the service that the custom domain applies to.
+ * @param {string} customDomainName the custom domain name.
+ */
+async function addCustomDomain(serviceARN, customDomainName) {
+    const data = await appRunnerClient.associateCustomDomain({
+        DomainName: customDomainName,
+        ServiceArn: serviceARN,
+    }).promise().catch(err => {
+        if (err.name === ERR_NAME_INVALID_REQUEST && err.message.includes(`${customDomainName} is already associated with`)) {
+            throw new CustomDomainError(`${customDomainName} is already associated with service ${serviceARN}`, ERR_NAME_CUSTOM_DOMAIN_ALREADY_ASSOCIATED);
+        }
+        throw err;
+    });
+
+    await upsertCNAMERecordAndWait(customDomainName, data.DNSTarget, appHostedZoneID);
+    await validateCertForDomain(serviceARN, customDomainName);
+}
+
+/**
+ * Validate certificates of the custom domain for the service by upserting validation records.
+ *
+ * @param {string} serviceARN ARN of the service that the custom domain applies to.
+ * @param {string} domainName the custom domain name.
+ * @throws wrapped error.
+ */
+async function validateCertForDomain(serviceARN, domainName) {
+    console.log("Add validation records");
+    let i;
+    for (i = 0; i < ATTEMPTS; i++){
+        const data = await appRunnerClient.describeCustomDomains({
+            ServiceArn: serviceARN,
+        }).promise().catch(err => {
+            throw new Error(`get custom domains for service ${serviceARN}: ` + err.message);
+        });
+
+        const customDomains = data.CustomDomains;
+        let domain;
+        for (const i in customDomains) {
+            if (customDomains[i].DomainName === domainName) {
+                domain = customDomains[i];
+                break;
+            }
+        }
+        if (!domain) {
+            throw new Error(`domain ${domainName} is not associated`);
+        }
+
+        if (domain.Status !== DOMAIN_STATUS_PENDING_VERIFICATION) {
+            console.log(`Custom domain ${domainName} status is ${domain.Status}. Desired state is ${DOMAIN_STATUS_PENDING_VERIFICATION}. Wait and check again`);
+            await sleep(3000);
+            continue;
+        }
+
+        const records = domain.CertificateValidationRecords;
+        for (const i in records) {
+            await upsertCNAMERecordAndWait(records[i].Name, records[i].Value, appHostedZoneID).catch(err => {
+                throw new Error("upsert certificate validation record: " + err.message);
+            });
```

Comment:
Why do we need to call this function again? isn't it done already in `addCustomDomain` and in `validateCertForDomain` we just need to wait until app runner picks up the changes

Reply:
Here is an explanation: https://github.com/aws/copilot-cli/pull/2507#discussion_r662515739!